### PR TITLE
nixos/caddy: don't set ExecReload if enableReload is disabled

### PR DESCRIPTION
**Pull request by flokli**
Otherwise, setting services.caddy.enableReload to false fails in a very bad fashion:

The reload command still gets executed, but fails:

```
Apr 26 21:23:01 n1-rk1 systemd[1]: Reloading Caddy...
Apr 26 21:23:01 n1-rk1 caddy[70793]: {"level":"info","ts":1714166581.733018,"msg":"using provided configuration","config_file":"/etc/caddy/caddy_config","config_adapter":"caddyfile"}
Apr 26 21:23:01 n1-rk1 caddy[70793]: {"level":"warn","ts":1714166581.7353032,"msg":"Caddyfile input is not formatted; run 'caddy fmt --overwrite' to fix inconsistencies","adapter":"caddyfile","file":"/etc/caddy/caddy_config","line":3}
Apr 26 21:23:01 n1-rk1 caddy[70793]: Error: sending configuration to instance: performing request: Post "http://localhost:2019/load": dial tcp [::1]:2019: connect: connection refused
Apr 26 21:23:01 n1-rk1 systemd[1]: caddy.service: Control process exited, code=exited, status=1/FAILURE
Apr 26 21:23:01 n1-rk1 systemd[1]: Reload failed for Caddy.
```

… and the server is not restarted either, as a ExecReload= command is specified.

Fix this, by only setting ExecReload if the reload exists.

The first empty string is still necessary to reset the old option.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/web-servers/caddy/default.nix
Comment (drupol) [inline nixos/modules/services/web-servers/caddy/default.nix:367]:
```suggestion
        ExecReload = lib.optionals cfg.enableReload [ ''${lib.getExe cfg.package} reload ${runOptions} --force'' ];
```

**Review by drupol — DISMISSED** [review]

File: nixos/modules/services/web-servers/caddy/default.nix
Comment (flokli) [reply nixos/modules/services/web-servers/caddy/default.nix:367]: No, the empty list item is explicitly there to clear the list from any previous entries (see systemd man pages).

If we apply the change you proposed, enableDefault set to false would make the config in the upstream unit file shine through.

Comment (drupol) [inline nixos/modules/services/web-servers/caddy/default.nix:367]:
```suggestion
        ExecReload = [ "" ] ++ lib.optionals cfg.enableReload [ ''${lib.getExe cfg.package} reload ${runOptions} --force'' ];
```

**Review by emilylange — APPROVED** [review]
Makes sense to me.

Thanks :)

Ref: #217536

**Comment by drupol** [comment]
> Makes sense to me.
> 
> Thanks :)
> 
> Ref: #217536

Please don't merge immediatelly, there's still one comment of mine that needs to be addressed.

**Comment by emilylange** [comment]
> Please don't merge immediatelly, there's still one comment of mine that needs to be addressed.

Your review is blocking, no worries :)

File: nixos/modules/services/web-servers/caddy/default.nix
Comment (jian-lin) [inline nixos/modules/services/web-servers/caddy/default.nix:366]: It is a mistake to add this comment in https://github.com/NixOS/nixpkgs/pull/217536. See https://github.com/NixOS/nixpkgs/pull/217536#discussion_r1175279412.

While it is not relevant to this PR, it would be good to remove this comment.

Comment (emilylange) [reply nixos/modules/services/web-servers/caddy/default.nix:366]: Yes, I considered pointing this out here as well (as part of my review).

But I eventually decided against doing so, because of an extremely fuzzy and far fetched interpretation of that comment:

> Validating the configuration before applying it ensures we’ll get a proper error that will be reported when switching to the configuration

being something like

> [Caddy] validat[es] the configuration before applying [, which] ensures we'll get a proper error that will be reported when switching to the configuration.

Caddy will never accept a broken config.

But yeah, that's really far fetched.

If you feel like updating the comment, please do :)

Comment (flokli) [reply nixos/modules/services/web-servers/caddy/default.nix:366]: Let's deal with this in a followup.

File: nixos/modules/services/web-servers/caddy/default.nix
Comment (flokli) [reply nixos/modules/services/web-servers/caddy/default.nix:367]: Still don't think we should use `lib.optionals` if it's just a single string (but `lib.optional`), but happy to do the `lib.getExe` suggestion.

Comment (drupol) [reply nixos/modules/services/web-servers/caddy/default.nix:367]: Technically both solutions are valid and this is mostly an opinionated change.

I see that many people are making mistakes in `nixpkgs`, mixing the use of `lib.optional` and `lib.optionals`. For all these reasons, I think using `lib.optionals` when dealing with an attributes which is a list makes more sense.

Also, using `lib.optionals` reduce the cognitive load for newcomers since the brackets needs to be explicitly added.

In any case, this won't prevent me from sleeping if you don't do the change, it's just based on my thin experience in the Nix world.

Comment (flokli) [reply nixos/modules/services/web-servers/caddy/default.nix:367]: Hmmh, I misinterpreted as a "request changes" type of review, because the old red review didn't get cleared.

I personally prefer using `lib.optional`. These are kinds of personal preferences, I usually leave it up to the authors of PRs.

Comment (flokli) [reply nixos/modules/services/web-servers/caddy/default.nix]: Updated.

**Comment by flokli** [comment]
@drupol is there anything left, or can this be merged?

**Comment by flokli** [comment]
Thanks!

**Pull request merged**
